src/custom-element.js

```diff
@@ -715,6 +790,95 @@ export function createAmpElementProto(win, name, implementationClass) {
     this.classList.toggle('amp-notsupported', state);
   };
 
+  /**
+   * Whether the loading can be shown for this element.
+   * @return {boolean}
+   * @private
+   */
+  ElementProto.isLoadingEnabled_ = function() {
+    // No loading indicator will be shown if either one of these
+    // conditions true:
+    // 1. `noloading` attribute is specified;
+    // 2. The element has not been whitelisted;
+    // 3. The element is too small or has not yet been measured;
+    // 4. The element has already been laid out;
+    // 5. The element is a `placeholder` or a `fallback`;
+    // 6. The element's layout is not a size-defining layout.
+    if (this.loadingDisabled_ === undefined) {
+      this.loadingDisabled_ = this.hasAttribute('noloading');
+    }
+    if (this.loadingDisabled_ ||
+            !isLoadingAllowed(this.tagName) ||
+            this.layoutWidth_ < MIN_WIDTH_FOR_LOADING_ ||
+            this.layoutCount_ > 0 ||
+            isInternalOrServiceNode(this) ||
+            !isLayoutSizeDefined(this.layout_)) {
+      return false;
+    }
+    return true;
+  };
+
+  /**
+   * Creates a loading object. The caller must ensure that loading can
+   * actually be shown. This method must also be called in the mutate
+   * context.
+   * @private
+   */
+  ElementProto.prepareLoading_ = function() {
+    if (!this.loadingContainer_) {
+      let container = document.createElement('div');
+      container.setAttribute('loading', '');
+      container.classList.add('-amp-loading-container');
+      container.classList.add('-amp-fill-content');
+      container.classList.add('amp-hidden');
+
+      let element = createLoaderElement();
+      container.appendChild(element);
+
+      this.appendChild(container);
+      this.loadingContainer_ = container;
+      this.loadingElement_ = element;
+    }
+  };
+
+  /**
+   * Turns the loading indicator on or off.
+   * @param {boolean} state
+   * @param {boolean=} opt_cleanup
+   * @private @final
+   */
+  ElementProto.toggleLoading_ = function(state, opt_cleanup) {
+    if (!state && !this.loadingContainer_) {
+      return;
+    }
+
+    // Check if loading should be shown.
+    if (state && !this.isLoadingEnabled_()) {
+      return;
+    }
+
+    vsync.mutate(() => {
+      if (state) {
+        this.prepareLoading_();
+      }
+      if (!this.loadingContainer_) {
+        return;
+      }
+
+      this.loadingContainer_.classList.toggle('amp-hidden', !state);
+      this.loadingElement_.classList.toggle('amp-active', state);
+
+      if (!state && opt_cleanup) {
+        let loadingContainer = this.loadingContainer_;
+        this.loadingContainer_ = undefined;
```

Comment:
Yeah. Ok. Switched to `null`. Seemed like `undefined` was warranted, but I agree.